drivers: retained_mem: ram: add support for nRF54L15

nika-nordic opened the pull request:
Add support for retained memory driver using nrfx_ram_ctrl helper.

hubertmis commented on samples/boards/nrf/system_off/src/retained_mem.c:
```diff
@@ -0,0 +1,77 @@
+
+const static struct device *retained_mem_device = DEVICE_DT_GET(DT_ALIAS(retainedmemdevice));
+
+struct retained_data retained;
```
Is there a reason for this variable to be in the global scope?

nika-nordic replied:
yes - this structure is referenced directly by sample `main.c` code: https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/boards/nrf/system_off/src/main.c#L40-L43 and this is a scheme enforced by existing POWER-based implementation for nRF52: https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/boards/nrf/system_off/src/retained.c

I would like to avoid major refactors of this sample, as the goal was to add support for nRF54L15 rather than making it "perfect"

nordic-krch commented on samples/boards/nrf/system_off/src/retained_mem.c:
```diff
@@ -0,0 +1,77 @@
+#define RETAINED_CRC_OFFSET offsetof(struct retained_data, crc)
+#define RETAINED_CHECKED_SIZE (RETAINED_CRC_OFFSET + sizeof(retained.crc))
+
+bool retained_validate(void)
```
this function looks very like https://github.com/zephyrproject-rtos/zephyr/blob/34e4edc28a3806885821d863cc3ebc9c60f5b817/samples/boards/nrf/system_off/src/retained.c#L139 Why not merge this into a single file and avoid redundancy?

masz-nordic replied:
I agree, the difference between those two is so small they should be merged.

nika-nordic replied:
fixed

nordic-krch commented on tests/drivers/retained_mem/api/testcase.yaml:
```diff
@@ -17,3 +17,10 @@ tests:
+  drivers.retained_mem.api.ram.nrf54l15pdk_nrf54l15_cpuapp:
+    platform_allow: nrf54l15pdk_nrf54l15_cpuapp
+    extra_args:
+      - DTC_OVERLAY_FILE="nrf54l15pdk_nrf54l15_cpuapp_ram.overlay"
```
why not remove `_ram` suffix, remove this  configuration and just add `nrf54l15pdk_nrf54l15_cpuapp` to `platform_allow` for `drivers.retained_mem.api.ram` case.

nika-nordic replied:
fixed

nordic-krch commented on drivers/retained_mem/Kconfig.nrf:
```diff
@@ -8,3 +8,10 @@ config RETAINED_MEM_NRF_GPREGRET
 	  register support.
+
+config RETAINED_MEM_NRF_RAM_CTRL
+	bool "nRF retention driver based on RAM control helper"
```
imo this is to detailed description. Maybe just `nRF RAM retention driver`? User does not know and does not care what is `RAM control helper`.

nika-nordic replied:
fixed

masz-nordic commented on samples/boards/nrf/system_off/boards/nrf54l15pdk_nrf54l15_cpuapp_retained_mem.overlay:
```diff
@@ -0,0 +1,22 @@
+	};
+};
+
+&cpuapp_sram {
+	reg = <0x20000000 DT_SIZE_K(184)>;
+	ranges = <0x0 0x20000000 0x2e000>;
+};
```
It would be useful to add a comment why this is needed.

nika-nordic replied:
fixed

nordicjm commented on drivers/retained_mem/retained_mem_nrf_ram_ctrl.c:
```diff
@@ -0,0 +1,37 @@
+	return 0;
+}
+
+SYS_INIT(retained_mem_nrf_init, POST_KERNEL, 0);
```
I would think this should be configured earlier

nika-nordic replied:
fixed

nika-nordic commented on drivers/retained_mem/retained_mem_nrf_ram_ctrl.c:
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2023, Nordic Semiconductor ASA
```
fixed

nika-nordic commented on samples/boards/nrf/system_off/README.rst:
```diff
@@ -12,16 +12,16 @@ deep sleep on Nordic platforms.
-in a checksummed data structure.  The POWER peripheral is configured to keep
-the containing RAM section powered while in system-off mode.
+This sample can also can demonstrate RAM retention. By selecting
+``CONFIG_APP_USE_NRF_RETENTION=y`` or ``CONFIG_APP_USE_RETAINED_MEM``
```
fixed

nordicjm commented on drivers/retained_mem/Kconfig.nrf:
```diff
@@ -8,3 +8,10 @@ config RETAINED_MEM_NRF_GPREGRET
+	bool "nRF RAM retention driver"
+	default y
+	depends on DT_HAS_ZEPHYR_RETAINED_RAM_ENABLED
+	help
```
this needs `depends on SOC_NRF54L15` or `depends on SOC_SERIES_NRF54LX` (whichever if applicable)

nordicjm replied:
Also needs to depend on `RETAINED_MEM_ZEPHYR_RAM`

nika-nordic replied:
done

nordicjm commented on drivers/retained_mem/Kconfig.nrf:
```diff
@@ -8,3 +8,11 @@ config RETAINED_MEM_NRF_GPREGRET
+	bool "nRF RAM retention driver"
+	default y
+	depends on DT_HAS_ZEPHYR_RETAINED_RAM_ENABLED && RETAINED_MEM_ZEPHYR_RAM
+	depends on SOC_SERIES_NRF52X || SOC_SERIES_NRF53X || SOC_SERIES_NRF54LX || SOC_SERIES_NRF91X
```
`depends on SOC_SERIES_NRF52X || SOC_SERIES_NRF53X || SOC_SERIES_NRF54LX || SOC_SERIES_NRF91X`

What? Why is this needed on anything that isn't nrf54l?

nika-nordic replied:
Purpose of this PR was to mainly support nrf54l, however the updated `retained_mem` driver supports nrf52/nrf53/nrf91 as well, hence this list of ORed SoC series

nordicjm replied:
Support how? The RAM is retained by default which is what I assume `nrfx_ram_ctrl_retention_enable_set` sets, so why is that needed when it's already enabled?

nordicjm replied:
All RAM is enabled by default, I don't mean retention I just mean RAM itself, which means you can reboot an application and keep the contents, this definitely works because we have 2 tests that check it on real hardware in https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/boot/mcuboot_data_sharing and https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/boot/mcuboot_recovery_retention

So this allows retaining RAM contents if someone turns the RAM section off... but why? If someone turns RAM sections off, they should know what they are doing, and having a part that is retained that you can't do anything with makes no sense to me. Not to mention that as I found out and posted on NCSDK-22549 after a gruelling battle with an ncs sample application and debugger, turning off RAM sections is a great way to lock the whole CPU up - or brick the module.

hubertmis replied:
This is about preserving partial RAM content when using https://docs.zephyrproject.org/apidoc/latest/group__sys__poweroff.html#gaf4552a8ea40b4adb94f85b0a1ff06e39

This function results in SystemOff state in nRF52/53/91 in which the RAM is powered down and its content is unretained by default.

nordicjm replied:
Interesting, then this should depends on `POWEROFF`

nika-nordic replied:
added dependency on `POWEROFF`

nordicjm replied:
you have added the wrong dependency, it needs to be `POWEROFF` not `HAS_POWEROFF`

nika-nordic committed:
drivers: retained_mem: nrf: implement using nrfx_ram_ctrl helper

Now retained_mem driver can utilize nrfx_ram_ctrl helper
for nRF devices. It provides abstraction over various RAM retention
peripherals like POWER, VMC or MEMCONF.

Signed-off-by: Nikodem Kastelik <nikodem.kastelik@nordicsemi.no>

nika-nordic committed:
boards: nordic: nrf54l15pdk: add retained_mem to supported drivers

This driver is now supported on nRF54L15 device.

Signed-off-by: Nikodem Kastelik <nikodem.kastelik@nordicsemi.no>

nika-nordic committed:
tests: drivers: retained_mem: api: add nrf54l15 support

As nRF54L15 supports retained_mem driver,
this test should be executed on this target.

Signed-off-by: Nikodem Kastelik <nikodem.kastelik@nordicsemi.no>

nika-nordic committed:
samples: boards: nrf: system_off: add support for retained_mem

Now the sample can showcase RAM retention using retained_mem driver.

Signed-off-by: Nikodem Kastelik <nikodem.kastelik@nordicsemi.no>
